core/src/main/java/com/linecorp/armeria/common/util/DefaultAsyncLoader.java

```diff
@@ -0,0 +1,268 @@
+import java.util.function.Function;
+import java.util.function.Predicate;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.linecorp.armeria.common.annotation.Nullable;
+
+final class DefaultAsyncLoader<T> implements AsyncLoader<T> {
+
+    private static final Logger logger = LoggerFactory.getLogger(DefaultAsyncLoader.class);
+
+    @SuppressWarnings("rawtypes")
+    private static final AtomicReferenceFieldUpdater<
+            DefaultAsyncLoader, RefreshingFuture> loadFutureUpdater = AtomicReferenceFieldUpdater
+            .newUpdater(DefaultAsyncLoader.class, RefreshingFuture.class, "loadFuture");
+
+    private final Function<@Nullable T, CompletableFuture<T>> loader;
+    @Nullable
+    private final Duration expireAfterLoad;
+    @Nullable
+    private final Predicate<@Nullable T> expireIf;
+    @Nullable
+    private final Predicate<@Nullable T> refreshIf;
+    @Nullable
+    private final BiFunction<Throwable, @Nullable T, @Nullable CompletableFuture<T>> exceptionHandler;
+
+    private volatile RefreshingFuture<T> loadFuture = RefreshingFuture.completedFuture();
+
+    DefaultAsyncLoader(Function<@Nullable T, CompletableFuture<T>> loader,
+                       @Nullable Duration expireAfterLoad,
+                       @Nullable Predicate<@Nullable T> expireIf,
+                       @Nullable Predicate<@Nullable T> refreshIf,
+                       @Nullable BiFunction<
+                               Throwable, @Nullable T, @Nullable CompletableFuture<T>> exceptionHandler) {
+        requireNonNull(loader, "loader");
+        this.loader = loader;
+        this.expireAfterLoad = expireAfterLoad;
+        this.expireIf = expireIf;
+        this.refreshIf = refreshIf;
+        this.exceptionHandler = exceptionHandler;
+    }
+
+    @Override
+    public CompletableFuture<T> get() {
+        return maybeLoad().thenApply(f -> f.value);
+    }
+
+    private CompletableFuture<CacheEntry<T>> maybeLoad() {
+        RefreshingFuture<T> future;
+        CacheEntry<T> cacheEntry = null;
+        for (;;) {
+            final RefreshingFuture<T> loadFuture = this.loadFuture;
+            if (!loadFuture.isDone()) {
+                return loadFuture;
+            }
+
+            if (!loadFuture.isCompletedExceptionally()) {
+                final CacheEntry<T> cacheEntry0 = loadFuture.join();
+                if (isValid(cacheEntry0)) {
+                    cacheEntry = cacheEntry0;
+                    final boolean needsRefresh = needsRefresh(cacheEntry);
+                    if (!needsRefresh || loadFuture.refreshing) {
+                        if (needsRefresh) {
+                            logger.debug("Refreshing already in progress, " +
+                                    "so not refresh it again. cacheEntry {}", cacheEntry);
+                        }
+                        return loadFuture;
+                    }
+                    logger.debug("Refresh cacheEntry {}", cacheEntry);
```

Comment:
It would be enough to log when refreshing a new value.
```suggestion
                        return loadFuture;
                    }
                    logger.debug("Pre-fetching a new value. loader: {}, cache: {}", loader, cacheEntry.value);
```

Reply:
✅ updated